cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1451,6 +1451,23 @@ struct CaseWhenFunction : ScalarFunction {
     if (auto kernel = DispatchExactImpl(this, *types)) return kernel;
     return arrow::compute::detail::NoMatchingKernel(this, *types);
   }
+
+  static std::shared_ptr<MatchConstraint> DecimalMatchConstraint() {
+    static auto constraint =
+        MatchConstraint::Make([](const std::vector<TypeHolder>& types) -> bool {
+          DCHECK_GE(types.size(), 2);
+          DCHECK(std::all_of(types.begin() + 1, types.end(), [](const TypeHolder& type) {
+            return is_decimal(type.id());
+          }));
+          const auto& ty1 = checked_cast<const DecimalType&>(*types[1].type);
```

Comment:
Is the cast actually needed? The `Equals` method should work as well without it.

Reply:
Oh, good catch! I was casting then comparing precisions and scales. Then I found the `Equals()` method should work and forgot to remove the cast. Updated to use `TypeHolder`'s `operator ==`. Thanks.